### PR TITLE
feat: CGGMP21 Key Refresh and Key Rotation

### LINKED ISSUE #57: [TASK] Add Keyrefresh for CGGMP21 gadget
- [ ] Use Phase3 to do a new key refresh
- [ ] Use Phase4 for creating a new key with new set of participants.

### PR DESCRIPTION
**Pull request by shekohex**
Closes #57

### CODE REVIEW

File: protocols/dfns-cggmp21/src/lib.rs [hunk 1]
```diff
@@ -17,143 +19,102 @@ pub mod error;
...
-    fn logger(&self) -> DebugLogger {
-        self.logger.clone()
-    }
-
-    fn client(&self) -> Self::Client {
-        self.client.clone()
-    }
-}
+// A macro to declare all the protocols
+decl_porto!(
```
Comment (drewstone) [inline protocols/dfns-cggmp21/src/lib.rs:96]: Nice!

File: protocols/dfns-cggmp21/src/lib.rs [hunk 2]
```diff
@@ -17,143 +19,102 @@ pub mod error;
...
-    pub network: N,
-    pub keystore_backend: ECDSAKeyStore<KBE>,
-    pub client: C,
-    pub logger: DebugLogger,
-    pub pallet_tx: Arc<dyn PalletSubmitter>,
-    pub _pd: std::marker::PhantomData<(B, BE)>,
-}
+/// A Helper macro to declare a protocol, used
+/// to avoid code duplication.
+macro_rules! decl_porto {
```
Comment (drewstone) [inline protocols/dfns-cggmp21/src/lib.rs:24]: Should this be abstracted into common macros? Usable in other protocols? CC @tbraun96

**Review by drewstone — APPROVED** [review]
Looks good, I wonder if we should make the following nit/fixes:
- `key_rotate` and `key_refresh` (in file names and variable naming since we are camel casing these two).
- Abstracting macros to be resuable and updating other protocols.
- Generalizing process for key rotation so that there's resuable code across all DKG protocols (it's basically the same process as run Phase1 and create linked records for any protocol).

**Comment by drewstone** [comment]
@shekohex if you agree can you make those changes?

**Comment by shekohex** [comment]
> @shekohex if you agree w/ renaming can you make those changes?

Yes, working on it right now.

**Pull request merged**